### PR TITLE
Add game and deal datamodel

### PR DESCRIPTION
**Pull request by RoachOSL**
As in title but I also added saving to file feature and game lookup functionality.

### CODE REVIEW

File: src/main/java/dev/Roach/GameLookup.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import dev.Roach.datamodel.game.Game;
+import dev.Roach.datamodel.game.GamePojo;
+import dev.Roach.datamodel.gameLookup.GameDealResponse;
+import dev.Roach.fetchers.GamesFetcher;
+
+import java.net.http.HttpClient;
+import java.util.ArrayList;
+
+public class GameLookup {
+    HttpClient client = HttpClient.newBuilder().build();
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/GameLookup.java]: private final

File: src/main/java/dev/Roach/GameLookup.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    GamesFetcher gamesFetcher = new GamesFetcher(client);
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/GameLookup.java]: private final

File: src/main/java/dev/Roach/GameLookup.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+
+    public void giveTitleToGetListOFDealsWithStores(String gameTitle) {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/GameLookup.java]: This method should be returning proper datamodel information rather than print it

File: src/main/java/dev/Roach/fetchers/DealsFetcher.java [hunk 4]
```diff
@@ -19,16 +26,18 @@ public DealsFetcher(HttpClient client) {
...
-    public List<String> getAllDeals() {
+    public ArrayList<DealAllListPojo> getAllDeals() {
 
-        List<String> allPages = new ArrayList<>();
+        ArrayList<DealAllListPojo> allPages = new ArrayList<>();
+        JSONMapper jsonMapper = new JSONMapper();
+        ObjectMapper objectMapper = new ObjectMapper();
 
         HttpRequest initialRequest = HttpRequest.newBuilder()
                 .uri(URI.create("https://www.cheapshark.com/api/1.0/deals?pageNumber=0"))
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/DealsFetcher.java]: This URI could be made into a static field and be more descriptive

File: src/main/java/dev/Roach/fetchers/DealsFetcher.java [hunk 5]
```diff
@@ -19,16 +26,18 @@ public DealsFetcher(HttpClient client) {
         this.client = client;
     }
 
-    public List<String> getAllDeals() {
+    public ArrayList<DealAllListPojo> getAllDeals() {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/DealsFetcher.java]: Why the change to ArrayList?

File: src/main/java/dev/Roach/fetchers/GamesFetcher.java [hunk 6]
```diff
@@ -15,7 +20,9 @@ public GamesFetcher(HttpClient client) {
         this.client = client;
     }
 
-    public String getGameContainingKeyword(String keyword) {
+    public ArrayList<GamePojo> getGameContainingKeyword(String keyword) {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/GamesFetcher.java]: Why arraylist instead of list?

File: src/main/java/dev/Roach/fetchers/GamesFetcher.java [hunk 7]
```diff
@@ -53,12 +61,33 @@ public String getGameUsingID(int id) {
...
 
         } catch (InterruptedException | IOException exception) {
             exception.printStackTrace();
             return "";
         }
     }
+
+    public GameDealResponse getGameDealObjectUsingID(int id) {
+
+        JSONMapper jsonMapper = new JSONMapper();
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/GamesFetcher.java]: jsonMapper could be a class field

File: src/main/java/dev/Roach/fetchers/GamesFetcher.java [hunk 8]
```diff
@@ -53,12 +61,33 @@ public String getGameUsingID(int id) {
...
+
+        try {
+            HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create("https://www.cheapshark.com/api/1.0/games?id=" + id))
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/GamesFetcher.java]: Static part of the URI could be a static field

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.ArrayList;
+
+public class StoresFetcher {
+    private final HttpClient client;
+
+    public StoresFetcher(HttpClient client) {
+        this.client = client;
+    }
+
+    public ArrayList<StoreAllPojo> getAllShops() {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/StoresFetcher.java]: Why arrayList?

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+        JSONMapper jsonMapper = new JSONMapper();
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        try (FileWriter fw = new FileWriter("dataFromApi/ShopList.txt")) {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/StoresFetcher.java]: file location can be a static field

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create("https://www.cheapshark.com/api/1.0/stores"))
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/StoresFetcher.java]: URI could be a static field

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+        } catch (InterruptedException | IOException exception) {
+            exception.printStackTrace();
+            return new ArrayList<>();
+        }
+    }
+
+    public ArrayList<StoreAllPojo> readAllShopsFromFile() {
+
+        String filePath = "dataFromApi/ShopList.txt";
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/StoresFetcher.java]: filepath could be a static variable

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+        try {
+            if (new File(filePath).exists()) {
+                String json = new String(Files.readAllBytes(Paths.get(filePath)));
+                return objectMapper.readValue(json, new TypeReference<ArrayList<StoreAllPojo>>() {
+                });
+            } else {
+                return new ArrayList<>();
+            }
+        } catch (Exception e) {
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/StoresFetcher.java:69]: Why are you catching Exception (generic one) instead of specific?

File: src/test/java/GameLookupTest.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        System.setOut(new PrintStream(outContent));
+    }
+
+    @AfterEach
+    public void restoreStreams() {
+        System.setOut(originalOut);
+    }
+
+    @Test
+    public void testGiveTitleToGetListOFDealsWithStoresWithValidTitle() {
```
Comment (blazejprzyluski) [inline src/test/java/GameLookupTest.java:36]: You should be basing the test on the outcome object rather than what is present in the console

File: src/main/java/dev/Roach/fetchers/StoresFetcher.java (hunk 13 above)
Comment (RoachOSL) [reply src/main/java/dev/Roach/fetchers/StoresFetcher.java:69]: Didn't even notice that, changed to IOException.

File: src/main/java/dev/Roach/GameLookup.java [hunk 15]
```diff
@@ -4,47 +4,44 @@
...
                 break;
             }
         }
 
-        if (!gameFound) {
-            System.out.println("No exact match found for the title: " + gameTitle);
-            return;
+        if (!foundGame) {
+            System.out.println("No matches found for the title: " + gameTitle);
+            return new GameDealResponse();
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/GameLookup.java]: Keep in mind you will need to add handling of an empty object then

File: src/main/java/dev/Roach/fetchers/DealsFetcher.java [hunk 16]
```diff
@@ -16,27 +17,27 @@
...
+import java.util.List;
 
-
+@Setter
 public class DealsFetcher {
 
-    private final HttpClient client;
+    private HttpClient client = HttpClient.newBuilder().build();
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private static final String dealsApiUrl = "https://www.cheapshark.com/api/1.0/deals?";
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/fetchers/DealsFetcher.java]: upper case, same below (final variables can be lower case, but static final should be upper case)

File: src/main/java/dev/Roach/GameLookup.java (hunk 15 above)
Comment (RoachOSL) [reply src/main/java/dev/Roach/GameLookup.java]: Changed to throw exception "NoSuchElementException" is this better approach or should I think about something else ? Also added isEmpty method in GameDealResponse class to make it work.

File: src/main/java/dev/Roach/GameLookup.java [hunk 17]
```diff
@@ -38,10 +38,15 @@ public GameDealResponse giveTitleToGetListOFDealsWithStores(String gameTitle) {
...
-            System.out.println("No matches found for the title: " + gameTitle);
-            return new GameDealResponse();
+            throw new NoSuchElementException("No game deals found for the title: " + gameTitle);
         }
 
-        return gamesFetcher.getGameDealObjectUsingID(gameIdentificator);
+        GameDealResponse response = gamesFetcher.getGameDealObjectUsingID(gameIdentificator);
+
+        if (response.isEmpty()) {
+            throw new NoSuchElementException("GameDealResponse is empty for game ID: " + gameIdentificator);
```
Comment (blazejprzyluski) [inline src/main/java/dev/Roach/GameLookup.java]: Please make the message more user friendly, users will have no clue what GameDealResponse is

**Review by blazejprzyluski — APPROVED** [review]
Left a small comment, after this it will be ready to merge
